contract/contracts/forward_name_resolver/forward_name_resolver.ral

```diff
@@ -0,0 +1,259 @@
+    pub fn generateToken() -> ByteVec {
+        let tokenExists = contractExists!(subContractId!(getRewardTokenKey()))
+        assert!(!tokenExists, ErrorCodes.TokenAlreadyGenerated)
+        let caller = callerAddress!()
+
+        let (encodeImmutableFields, encodeMutableFields) = RewardToken.encodeFields!(
+            selfContractId!(),
+            b`ANS`,
+            b`Alephium Name Service`,
+            18,
+            0
+        )
+
+        return copyCreateSubContractWithToken!{caller -> ALPH: minimalContractDeposit!()}(
+            getRewardTokenKey(),
+            tokenTemplateId,
+            encodeImmutableFields,
+            encodeMutableFields,
+            u256Max!()
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    fn createName(minter: Address, index: U256, name: ByteVec, capitalisation: ByteVec) -> ByteVec {
+        let now = blockTimeStamp!()
+        let expires = now + renewLength
+        let (encodeImmutableFields, encodeMutableFields) = Name.encodeFields!(
+            selfContractId!(),  // <- collectionId
+            index,              // <- nftIndex
+            name,               // <- name
+            capitalisation,     // <- capitalisation
+            minter,             // <- initialAddress
+            expires             // <- expires
+        )
+
+        let newContractId = copyCreateSubContractWithToken!{minter -> ALPH: minimalContractDeposit!()}(
+            getNftKey(index),
+            nameTemplateId,
+            encodeImmutableFields,
+            encodeMutableFields,
+            1,
+            minter
+        )
+
+        emit NameCreated(
+            index,
+            name,
+            capitalisation,
+            minter,
+            expires
+        )
+
+        return newContractId
+    }
+
+    @using(preapprovedAssets = true)
+    fn handleRewardTokens(caller: Address, name: ByteVec) -> () {
+        let nameLength = size!(name)
+        let rewardToken = getRewardToken()
+        let rewardTokenId = tokenId!(rewardToken)
+
+        if (nameLength > 8) {
+            rewardToken.mintTokens(caller, 1 alph)
+        }
+        if (nameLength <= 8) {
+            let cost = (9 - nameLength) * 100
+            rewardToken.burnTokens{caller -> rewardTokenId: cost * 1 alph}(caller, cost * 1 alph)
+        }
+    }
+
+    @using(preapprovedAssets = true, updateFields = true, checkExternalCaller = false)
+    pub fn mint(name: ByteVec, capitalisation: ByteVec) -> ByteVec {
+        let caller = callerAddress!()
+        totalSupply = totalSupply + 1
+        let rewardToken = getRewardToken()
+        let rewardTokenId = tokenId!(rewardToken)
+        
+        if (nameNftIndex.contains!(name)) {
+            let existingNft = getNftByName(name)
+            let expired = hasExpired(existingNft)
+            assert!(expired, ErrorCodes.NameHasNotExpired)
+            existingNft.delete(caller)
+            nameNftIndex[name] = totalSupply
```

Comment:
It seems this would prevent the previous name owner from redeeming the ALPH deposited when creating the map entry.

For example, if user A first minted the name `foo.alph`, and after it expires, user B mints the name `foo.alph`, the `nameNftIndex` map will be updated, and user B won't pay the fee for the map entry. User A will fail to call `deleteExpired` because the name no longer belongs to user A. However, when user B calls `deleteExpired`, the map entry will be deleted from the `nameNftIndex`, and the deposited ALPH will be transferred to user B, even though the deposited ALPH should belong to user A.

Reply:
The `mint` function will actually delete the previous NFT contract so user A won't be able to call `deleteExpired`.

Instead they will have to burn the token.
I think its fine if the next owner gets to keep the old deposit?

Reply:
I've made `deleteExpired` public so the previous comment no longer applies.